content/02-model/01-train-and-deploy-model.qmd

```diff
@@ -0,0 +1,300 @@
+
+```{r fit_model}
+model_spec <- 
+  rand_forest() %>%
+  set_mode("regression") %>%
+  set_engine("ranger")
+
+model_workflow <- 
+  workflow() %>%
+  add_recipe(recipe_spec) %>%
+  add_model(model_spec)
+
+model_fit <- fit(model_workflow, data = train_data)
+model_fit
+```
+
+## Model evaluation
+
+```{r evaluate_model}
+predictions <- predict(model_fit, test_data)
+
+results <- test_data %>%
+  mutate(preds = predictions$.pred)
+
+oos_metrics(results$n_bikes, results$preds)
+```
+
+## Model deployment
+
+### `vetiver`
+
+Create a `vetiver` model object.
+
+```{r create_vetiver_model}
+model_name <- "bike_predict_model_r"
+pin_name <- glue("sam.edwardes/{model_name}")
+
+v <- vetiver_model(
+  model_fit, 
+  model_name,
+  versioned = TRUE,
+  save_ptype = train_data %>%
+    head(1) %>%
+    select(-n_bikes),
+  metadata = list(
+    train_dates = c(min(train_data$date), max(train_data$date)),
+    test_dates = c(min(test_data$date), max(test_data$date))
+  )
+)
+
+v
+```
+
+### `pins`
+
+Save the model as a *pin* to RStudio Connect:
+
+```{r pin_model}
+# Use RStudio Connect as a board.
+board <- pins::board_rsconnect(
+  server = Sys.getenv("CONNECT_SERVER"),
+  key = Sys.getenv("CONNECT_API_KEY"),
+  versioned = TRUE
+)
+
+# Write the model to the board.
+board %>%
+ vetiver_pin_write(vetiver_model = v)
+```
+
+Update the pin permissions and metadata on Connect:
+
+```{r pin_metadata}
+# Set content permissions
+client <- connectapi::connect(
+  server = Sys.getenv("CONNECT_SERVER"),
+  api_key = Sys.getenv("CONNECT_API_KEY")
+)
+
+# Get the content
+content <- client %>%
+  connectapi::content_item("0273207f-0f1f-44cb-845e-408f9dae2dd6")
+
+content$update(title = "Bike Predict (R) - Model - Pinned Model")
+
+# Make the app visible to the world!
+content$update(access_type = "all")
+
+# Define the contents collaborators
+collaborators <- c(
+  "69db516b-ba3e-442a-84d9-f916fc92a2ca", # Gagan Singh
+  "99a43fac-2edd-4c31-8cba-a001c507da2e"  # Xu Fei
+  
+)
+for (c in collaborators) {
+  content %>%
+    connectapi::content_add_user(c, role = "owner")
+}
+# Get tags.
+tags <- client %>%
+  connectapi::get_tags()
+  
+  # Update content settings.
+content %>%
+  connectapi::set_content_tags(tags$`Projects and Presentations`$`Bike Predict`$R$Model) %>%
+  connectapi::set_image_url("https://i.imgur.com/ITQOzRw.png") %>%
+  connectapi::set_vanity_url("bike-predict-r-pinned-model", force = TRUE)
+```
+
+### `plumber`
+
+Convert the model into a plumber API. The function `vetiver_write_plumber` will generate the plumber code for you and write it to *plumber.R*.
+
+```{r write_model_to_plumber}
+# Write the model to `api/plumber.R`.
+dir.create("api")
+vetiver_write_plumber(board, pin_name, file = "api/plumber.R")
+
+# Write a manifest.json file for the api
+rsconnect::writeManifest("api")
+```
+
+Here is what the plumber file looks like:
+
+```{r preview_plumber_file}
+readLines("api/plumber.R") %>%
+  paste(collapse = "\n") %>%
+  cat()
+```
```

Comment:
is there a reason the `plumber.R` file is not included? We should have one version of how this file looks like in the repo.

Reply:
I chose not to include `plumber.R` because it is an auto generated bit of code and should not be edited by hand. I do print the contents of `plumber.R` just to demonstrate to readers what the output of `vetiver::vetiver_write_plumber` is.

Reply:
Updated code to keep api/plumber.R in source code: d398a51.